File: captainhcg/py/41-first-missing-positive.py
class Solution(object):
    def firstMissingPositive(self, nums):
        idx, length = 0, len(nums)
        while idx < length:
            n = nums[idx]
            if n <= 0 or n > length or n - 1 == idx:
                pass
            else:
                should_be_at = n - 1
                if nums[idx] != nums[should_be_at]:
                    nums[should_be_at], nums[idx] = nums[idx], nums[should_be_at]
                    continue
            idx += 1

        for idx, n in enumerate(nums):
            if n - 1 != idx:
                return idx + 1

        return length + 1

File: captainhcg/go/41-first-missing-positive.go
func firstMissingPositive(nums []int) int {
    i := 0
    for i < len(nums){
        value := nums[i]
        if value > 0 && value <= len(nums){
            should_be_at := value - 1
            if value != nums[should_be_at]{
                nums[should_be_at], nums[i] = nums[i], nums[should_be_at]
                continue
            }
        }
        i += 1
    }
    
    for idx, v := range(nums){
        if v != idx + 1{
            return idx + 1
        }
    }

    return len(nums) + 1
}
